[Cocoa] Add conversion functions from native modifier types to OptionSet<WebKit::WebEventModifier>

ddkilzer opened the pull request:
#### 8996a7f84e0fde0d7b15e727404daea7b8db8dc6
<pre>
[Cocoa] Add conversion functions from native modifier types to OptionSet&lt;WebKit::WebEventModifier&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=248162">https://bugs.webkit.org/show_bug.cgi?id=248162</a>
&lt;rdar://102571520&gt;

Reviewed by Darin Adler and Timothy Hatcher.

Define conversion functions from native modifier types to
OptionSet&lt;WebKit::WebEventModifier&gt; to make it easier to reuse
this code.

* Source/WebKit/Shared/ios/WebIOSEventFactory.h:
(WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags): Add.
* Source/WebKit/Shared/ios/WebIOSEventFactory.mm:
(WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags): Add.
- Move here from WKMouseGestureRecognizer.mm.

* Source/WebKit/Shared/mac/WebEventFactory.h:
(WebKit::WebEventFactory::webEventModifiersForNSEventModifierFlags): Add.
* Source/WebKit/Shared/mac/WebEventFactory.mm:
(WebKit::modifiersForEvent): Delete.
(WebKit::WebEventFactory::createWebMouseEvent):
(WebKit::WebEventFactory::createWebWheelEvent):
(WebKit::WebEventFactory::createWebKeyboardEvent):
- Switch to use webEventModifiersForNSEventModifierFlags().
(WebKit::WebEventFactory::webEventModifiersForNSEventModifierFlags): Add.
- Move and rename from modifiersForEvent().  Change argument
  type from native event to native modifier flags.

* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::WebAutomationSession::platformWebModifiersFromRaw):
- Remove duplicate code by calling
  WebEventFactory::webEventModifiersForNSEventModifierFlags().
* Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.mm:
(webEventModifiersForUIKeyModifierFlags): Delete.
- Move to WebIOSEventFactory.mm.
(-[WKMouseGestureRecognizer createMouseEventWithType:wasCancelled:]):
- Switch to use
  WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags().

Canonical link: <a href="https://commits.webkit.org/256942@main">https://commits.webkit.org/256942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc0076a62cceeac0779358c66bbb71753204b7b2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97317 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6581 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30473 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106836 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6882 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35318 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89727 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/103515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102980 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83946 "Built successfully") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/32167 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/87018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88843 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/597 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/580 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5380 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2352 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1824 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41107 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebKit/Shared/ios/WebIOSEventFactory.h:
```diff
@@ -44,6 +44,7 @@ class WebIOSEventFactory {
     static WebKit::WebWheelEvent createWebWheelEvent(UIScrollEvent *, UIView *contentView, std::optional<WebKit::WebWheelEvent::Phase> overridePhase = std::nullopt);
 #endif
 
+    static OptionSet<WebKit::WebEventModifier> fromUIKeyModifierFlags(UIKeyModifierFlags);
```
Strongly prefer function names that name the return type rather than the argument type. But maybe here `UIKeyModifierFlags` is just a name for an integer type?

ddkilzer replied:
`UIKeyModifierFlags` is an `NSInteger` (`long`) as far as C++ is concerned as it is defined as an `NS_OPTIONS()` enum.

Will rename the method back to `webEventModifiersForUIKeyModifierFlags()`.

darinadler commented on Source/WebKit/Shared/mac/WebEventFactory.h:
```diff
@@ -53,6 +53,7 @@ class WebEventFactory {
     static bool shouldBeHandledAsContextClick(const WebCore::PlatformMouseEvent&);
 
 #if defined(__OBJC__)
+    static OptionSet<WebKit::WebEventModifier> fromNSEventModifierFlags(NSEventModifierFlags);
```
Strongly prefer function names that name the return type rather than the argument type. But maybe here `NSEventModifierFlags` is just a name for an integer type?

ddkilzer replied:
`NSEventModifierFlags ` is an `NSUInteger` (`unsigned long`) as far as C++ is concerned as it is defined as an `NS_OPTIONS()` enum.

Will rename the method to `webEventModifiersForNSEventModifierFlags()` for consistency.

darinadler commented on Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
```diff
@@ -253,20 +254,7 @@ static WebMouseEventButton automationMouseButtonToPlatformMouseButton(MouseButto
 
 OptionSet<WebEventModifier> WebAutomationSession::platformWebModifiersFromRaw(unsigned modifiers)
```
Can we get rid of this function? Why do we need a platform-independent “raw”?

ddkilzer replied:
Yes, I think so.  I see it's private to `WebAutomationSession`, but I need to rework the GTK port as well:

```
Source/WebKit/Shared/gtk/WebEventFactory.cpp
Source/WebKit/UIProcess/Automation/gtk/WebAutomationSessionGtk.cpp
```

I could also just make `WebAutomationSession::platformWebModifiersFromRaw()` an inline method on macOS.

ddkilzer replied:
> I could also just make `WebAutomationSession::platformWebModifiersFromRaw()` an inline method on macOS.

To do this, I need to make `Source/WebKit/Shared/Mac/WebEventFactory.h` safe to include in C++ source.  The next patch will do this, so I want to wait on this change.

Commit:
[Cocoa] Add conversion functions from native modifier types to OptionSet<WebKit::WebEventModifier>
https://bugs.webkit.org/show_bug.cgi?id=248162
<rdar://102571520>

Reviewed by Darin Adler and Timothy Hatcher.

Define conversion functions from native modifier types to
OptionSet<WebKit::WebEventModifier> to make it easier to reuse
this code.

* Source/WebKit/Shared/ios/WebIOSEventFactory.h:
(WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags): Add.
* Source/WebKit/Shared/ios/WebIOSEventFactory.mm:
(WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags): Add.
- Move here from WKMouseGestureRecognizer.mm.

* Source/WebKit/Shared/mac/WebEventFactory.h:
(WebKit::WebEventFactory::webEventModifiersForNSEventModifierFlags): Add.
* Source/WebKit/Shared/mac/WebEventFactory.mm:
(WebKit::modifiersForEvent): Delete.
(WebKit::WebEventFactory::createWebMouseEvent):
(WebKit::WebEventFactory::createWebWheelEvent):
(WebKit::WebEventFactory::createWebKeyboardEvent):
- Switch to use webEventModifiersForNSEventModifierFlags().
(WebKit::WebEventFactory::webEventModifiersForNSEventModifierFlags): Add.
- Move and rename from modifiersForEvent().  Change argument
  type from native event to native modifier flags.

* Source/WebKit/UIProcess/Automation/mac/WebAutomationSessionMac.mm:
(WebKit::WebAutomationSession::platformWebModifiersFromRaw):
- Remove duplicate code by calling
  WebEventFactory::webEventModifiersForNSEventModifierFlags().
* Source/WebKit/UIProcess/ios/WKMouseGestureRecognizer.mm:
(webEventModifiersForUIKeyModifierFlags): Delete.
- Move to WebIOSEventFactory.mm.
(-[WKMouseGestureRecognizer createMouseEventWithType:wasCancelled:]):
- Switch to use
  WebIOSEventFactory::webEventModifiersForUIKeyModifierFlags().

Canonical link: https://commits.webkit.org/256942@main
